subxt/src/backend/utils.rs

```diff
@@ -0,0 +1,293 @@
+            Self::Resubscribe { .. } => write!(f, "Resubscribe"),
+            Self::Closed(_) => write!(f, "Closed"),
+        }
+    }
+}
+
+/// Retry subscription.
+struct RetrySubscription<T> {
+    resubscribe: ResubscribeGetter<T>,
+    state: PendingOrStream<T>,
+}
+
+impl<T> std::marker::Unpin for RetrySubscription<T> {}
+
+impl<T> Stream for RetrySubscription<T> {
+    type Item = Result<T, Error>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        loop {
+            let next_state = match self.state {
+                PendingOrStream::Stream(ref mut s) => match ready!(s.poll_next_unpin(cx)) {
+                    Some(Err(err)) => {
+                        if !err.is_disconnected_will_reconnect() {
+                            NextState::Closed(Some(err))
+                        } else {
+                            NextState::Resubscribe {
+                                pending: (self.resubscribe)(),
+                                err,
+                            }
+                        }
+                    }
+                    None => NextState::Closed(None),
+                    Some(Ok(val)) => return Poll::Ready(Some(Ok(val))),
+                },
+                PendingOrStream::Pending(ref mut fut) => match ready!(fut.poll_unpin(cx)) {
+                    Ok(stream) => NextState::Stream(stream),
+                    Err(err) => {
+                        if !err.is_disconnected_will_reconnect() {
+                            NextState::Closed(Some(err))
+                        } else {
+                            NextState::Resubscribe {
+                                pending: (self.resubscribe)(),
+                                err,
+                            }
+                        }
+                    }
+                },
+                PendingOrStream::Closed => NextState::Closed(None),
+            };
+
+            match next_state {
+                NextState::Resubscribe { pending, err } => {
+                    self.state = PendingOrStream::Pending(pending);
+                    return Poll::Ready(Some(Err(err)));
+                }
+                NextState::Stream(stream) => {
+                    self.state = PendingOrStream::Stream(stream);
+                }
+                NextState::Closed(maybe_err) => {
+                    self.state = PendingOrStream::Closed;
+
+                    if let Some(err) = maybe_err {
+                        return Poll::Ready(Some(Err(err)));
+                    } else {
+                        return Poll::Ready(None);
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Retry a future until it doesn't return a disconnected error.
+pub async fn retry<T, F, R>(mut retry_future: F) -> Result<R, Error>
+where
+    F: FnMut() -> T,
+    T: Future<Output = Result<R, Error>>,
+{
+    const REJECTED_MAX_RETRIES: usize = 10;
```

Comment:
It's hard to reproduce now and I think I managed to produce once so far after the changes to clean the state after `DisconnectWillReconnect` occured.

One retry was sufficient then and 10 is probably not needed.

Reply:
I'm not sure that this is an issue with the RPC APIs themselves. We can raise a separate issue I think to tweak the code such that we are guaranteed to get a valid subscription ID from the follow subscription once we detect a disconnect (which could come from a separate RPC call and not be known yet to the follow subscription task).

Reply:
Hmm, to me it's not really issue in the spec because the server should really return a valid subscription ID if it's made on a new connection. This is just relevant to "poll order" of the pending futures when reconnect occurs which we could fix i.e, OrderedFutures or something.

I think we could just do the following:


```rust
    let mut sub_id = get_subscription_id(&self.follow_handle).await?;
    
    loop {
    
        tokio::select!( {
            // probably need to add this as internal thing to backend
            _ = client.reconnected() => {
                sub_id = get_subscription_id(&self.follow_handle).await?;
            }
            res = retry(|| self.methods.chainhead_v1_header(&sub_id, at)) => {
                break res;
            }
        }
    }
```

Reply:
> This is just relevant to "poll order" of the pending futures when reconnect occurs

Yeah; basically:

- The follow stream subscription that you pull events from is driven in some task by some driver somewhere.
- RPC requests like `chainhead_v1_header().await` are driven (probably in a separate task).

So if the latter returns `DisconnectedWillReconnect`, it's no guarantee that the former has been polled and also has seen this error and handled it.

> I think we could just do the following

You _could_ add something like `.reconnected()` to the backend, but that doesn't necessarily mean that everything has been polled and synced?

One way to sync the two things might be to share some state between them, but I have a feeling that would become messy.

A way I have in mind to try out at some point is:

- maintain a counter in the follow subscription thing which starts at 0.
- every time we get a new `initialized` event, increment this counter.
- now, if some call relying on the follow subscription eg `chainhead_v1_header` returns `DisconnectedWillReconnect`, we note the current counter value and keep trying the follow subscription again until we see that the counter increments (ie, it's been polled and has noticed that disconnect happens)
- when this happens, we know that it's reconnected and try the new subscription ID.  

(We could maybe also do this by just waiting for the subscription ID to change, rather than waiting for a counter to increment, but we don't have any guarantee about subscription Ids being random etc, so it wouldn't be very safe to do it this way)

Reply:
But yeah, as long as the current hack seems to work for now, I'm happy if we look into this as a followup because it'll need some thought and iteration to try it and, if it seems good, abstract over the idea nicely and avoid over complicated code :)

Reply:
> You could add something like .reconnected() to the backend, but that doesn't necessarily mean that everything has been polled and synced?

Yes, you are correct but both are needed. It's not elegant to retry each future in loop and if we could abort the pending futures when we know that the connection is lost and retry them after the chainHead_v1_follow has been established again.

Yes, let's create an issue after this PR instead :)